### PR TITLE
Fixed #30027 -- Errored out on Window function usage if unsupported.

### PR DESCRIPTION
**Pull request by charettes**
https://code.djangoproject.com/ticket/30027

### CODE REVIEW

**Comment by charettes** [comment]
buildbot, test on oracle.

File: tests/expressions_window/tests.py [hunk 1]
```diff
@@ -536,7 +536,7 @@ def test_multiple_ordering(self):
             ('Brown', 53000, 'Sales', datetime.date(2009, 9, 1), 108000),
         ], transform=lambda row: (row.name, row.salary, row.department, row.hire_date, row.sum))
 
-    @skipIf(connection.vendor == 'postgresql', 'n following/preceding not supported by PostgreSQL')
```
Comment (pope1ni) [inline tests/expressions_window/tests.py]: This restriction was lifted in PostgreSQL 11.

See https://modern-sql.com/blog/2019-01/sqlite-in-2018#window-functions

Comment (charettes) [reply tests/expressions_window/tests.py]: I know but the test fails on PostgreSQL 11 somehow and I didn't want to investigate more

Using django-docker-box

```
$ POSTGRES_VERSION=11 docker-compose run --rm postgres --noinput expressions_window
[snip]
AssertionError: 'PostgreSQL only supports UNBOUNDED together with PRECEDING and FOLLOWING.' not found in 'RANGE with offset PRECEDING/FOLLOWING is not supported for column type date and offset type integer\nLINE 1: ...e_date" ASC RANGE BETWEEN UNBOUNDED PRECEDING AND 1 FOLLOWIN...\n                                                             ^\nHINT:  Cast the offset value to an appropriate type.\n'
```

It looks like there might be a bug in PostgreSQL 11 actually.

Anyway, this should be handled in a different PR.

**Comment by timgraham** [comment]
The test failure from the Windows build (which uses SQLite 3.26.0) seems to be non-determinstic (passing sometimes):
```
Traceback (most recent call last):
  File "C:\Jenkins\workspace\pull-requests-windows\database\sqlite3\label\windows\python\Python37\tests\expressions_window\tests.py", line 614, in test_subquery_row_range_rank
    ], transform=lambda row: (row.name, row.department, row.highest_avg_salary_date))
  File "C:\Jenkins\workspace\pull-requests-windows\database\sqlite3\label\windows\python\Python37\django\test\testcases.py", line 1055, in assertQuerysetEqual
    return self.assertEqual(list(items), values, msg=msg)
AssertionError: Lists differ: [('Ad[15 chars]ng', None), ('Jenson', 'Accounting', None), ('[308 chars]one)] != [('Ad[15 chars]ng', datetime.date(2005, 11, 1)), ('Jenson', '[544 chars] 1))]

First differing element 0:
('Adams', 'Accounting', None)
('Adams', 'Accounting', datetime.date(2005, 11, 1))
```

**Comment by charettes** [comment]
Looks like this might be related to this issue

_Window function in correlated subquery causes assertion fault_

https://www.sqlite.org/src/tktview/d0866b26f83e9c55e30d

http://sqlite.1065341.n5.nabble.com/Bug-in-3-25-2-RECURSIVE-CTE-window-function-td103958.html#a105246

I guess we'll have to wait for SQLite 3.26.1 release to confirm it's working appropriately.

**Comment by timgraham** [comment]
It seems to me that we need to wait to merge the `NotSupportedError` commit until the "Enabled window function support" commit, otherwise Windows functions will be prohibited, even on supported versions of SQLite.

**Comment by timgraham** [comment]
The problematic test passes 200 runs on SQLite 3.27 so I think this is good to go. I added a skip for older versions.

**Comment by charettes** [comment]
Awesome, thanks @timgraham!

**Pull request merged**